aptos-move/block-executor/src/executor.rs

```diff
@@ -121,8 +127,17 @@ impl<
                             // eventually finish and lead to unblocking txn_idx, contradiction.
                             let (lock, cvar) = &*dep_condition;
                             let mut dep_resolved = lock.lock();
-                            while !*dep_resolved {
+                            // dep_resolved.0 indicates whether the read dependency is resolved.
+                            while !(*dep_resolved).0 {
+                                // If the PE halts, read should return immediately.
```

Comment:
We need a longer comment here explaining why we are safe:
// When we early terminate, done is set first, and then threads waiting on dependencies are notified. By checking done here, we ensure this isn't a condvar due to stale state in the multi-version data-structure that will never be notified. If terminate had already passed the index of the current thread when sheduler.wait_for_dependency set the status, then done was set even earlier, and it will be observed. Otherwise, the status will be observed by the terminating thread and condvar will be notified.

Thinking about this, there might be a better way to structure the code:
scheduler.wait_for_dependency can return an enum:
Dependency(dep_condition)
Resolved, (same as none now)
SchedulerTerminated (or execution halted or smt)
Or it can just be an Err<Option< where error indicates that we are done.
Then we can match this condition here and return PEError (whatever we rename this to).

The actual logic will be done in scheduler, which is a much nicer place for it. will add a comment there.